src/internal/storage/fileset/compaction.go

```diff
@@ -93,53 +97,68 @@ func NewDistributedCompactor(s *Storage, maxFanIn int, workerFunc CompactionBatc
 
 // Compact compacts the contents of ids into a new fileset with the specified ttl and returns the ID.
 // Compact always returns the ID of a primitive fileset.
-func (c *DistributedCompactor) Compact(ctx context.Context, ids []ID, ttl time.Duration) (*ID, error) {
+func (c *DistributedCompactor) Compact(ctx context.Context, ids []ID, ttl time.Duration) (_ *ID, retErr error) {
 	var err error
 	ids, err = c.s.Flatten(ctx, ids)
 	if err != nil {
 		return nil, err
 	}
-	if len(ids) <= c.maxFanIn {
-		return c.shardedCompact(ctx, ids, ttl)
-	}
-	childSize := c.maxFanIn
-	for len(ids)/childSize > c.maxFanIn {
-		childSize *= c.maxFanIn
-	}
-	results := []ID{}
-	for start := 0; start < len(ids); start += childSize {
-		end := start + childSize
-		if end > len(ids) {
-			end = len(ids)
+	var id *ID
+	if err := c.s.WithRenewer(ctx, ttl, func(ctx context.Context, renewer *Renewer) error {
+		var taskLens []int
+		var tasks []CompactionTask
+		for start := 0; start < len(ids); start += c.maxFanIn {
+			end := start + c.maxFanIn
+			if end > len(ids) {
+				end = len(ids)
+			}
+			ids := ids[start:end]
+			taskLen := len(tasks)
+			if err := miscutil.LogStep(fmt.Sprintf("sharding %v file sets", len(ids)), func() error {
+				return c.s.Shard(ctx, ids, func(pathRange *index.PathRange) error {
+					tasks = append(tasks, CompactionTask{
+						Inputs:    ids,
+						PathRange: pathRange,
+					})
+					return nil
+				})
+			}); err != nil {
+				return err
+			}
+			taskLens = append(taskLens, len(tasks)-taskLen)
```

Comment:
I think this line and above can be outside the renewer?

Reply:
Yeah all of this just operates on that input, which we can assume will not expire.

Reply:
Yeah, will do.